phdi/fhir/tabulation/tables.py

```diff
@@ -166,26 +320,165 @@ def generate_all_tables_in_schema(
+        generate_table(schema, output_path, output_format, fhir_url, cred_manager)
 
 
 @cache
 def _get_fhirpathpy_parser(fhirpath_expression: str) -> Callable:
     """
-    Accepts a FHIRPath expression, and returns a callable function which returns the
-    evaluated value at fhirpath_expression for a specified FHIR resource.
+    Accepts a FHIRPath expression, and returns a callable function
+    which returns the evaluated value at fhirpath_expression for
+    a specified FHIR resource.
 
     :param fhirpath_expression: The FHIRPath expression to evaluate.
-    :return: A function that, when called passing in a FHIR resource, will return value
-      at `fhirpath_expression`.
+    :return: A function that, when called passing in a FHIR resource,
+      will return value at `fhirpath_expression`.
     """
     return fhirpathpy.compile(fhirpath_expression)
 
 
+def _get_reference_directions(schema: dict) -> dict:
+    """
+    Creates a dictionary mapping indicating how the resources that
+    will be used in creating the final output tables relate to each
+    other. For any column desired in an output table, it is possible
+    for the column to be found in a resource that either a) references
+    a given resource, or b) is referenced *by* the given resource.
+    Since each table in the schema is defined with an  "anchor" resource
+    (the main type of resource determining the number of rows in the
+    table), referenced resources of type A can be labeled "backward"
+    pointers and referenced resources of type B can be labeled "forward"
+    pointers. This mapping is used to efficiently group and aggregate
+    related resource data for tabulation.
+
+    :param schema: A user-defined schema featuring some number
+      (possibly 0) of indirectly referenced resources.
+    :return: A dictionary containing mappings, for each table, of
+      how referenced resources relate to the anchor resource.
+    """
+
+    directions_by_table = {}
+    for table_name, table_params in schema.get("tables", {}).items():
+        anchor_type = table_params.get("resource_type", "")
+        directions_by_table[table_name] = {
+            "anchor": anchor_type,
+            "forward": set(),
+            "reverse": {},
+        }
+
+        for _, column_params in table_params.get("columns", {}).items():
+            if "reference_location" in column_params:
+                [direction, ref_path] = column_params.get("reference_location").split(
+                    ":", 1
+                )
+                referenced_resource_type = column_params.get("fhir_path").split(".")[0]
+                if direction == "forward":
+                    directions_by_table[table_name][direction].add(
+                        referenced_resource_type
+                    )
+                else:
+                    directions_by_table[table_name][direction][
+                        referenced_resource_type
+                    ] = ref_path
```

Comment:
edge case - what if the same resource type is referenced twice from the same resource type? For instance:

```yaml
resource_type: Practitioner
columns:
  Interpreted Reports:
    fhir_path: DiagnosticReport.code.text
    reference_location: reverse:DiagnosticReport:resultsInterpreter
  Performed Reports: 
    fhir_path: DiagnosticReport.code.text
    reference_location: reverse:DiagnosticReport:performer
```

Thinking maybe you could make this element reference an array of `ref_path`s?
```suggestion
                    if directions_by_table[table_name][direction].get(referenced_resource_type) is None:
                        directions_by_table[table_name][direction][referenced_resource_type] = []
                    directions_by_table[table_name][direction][
                        referenced_resource_type
                    ].append(ref_path)
```

Reply:
The code can actually already account for this! The part that you've commented on is _just_ responsible for creating the directions of association between the various resource types. If one resource references an anchor multiple times (in the way you illustrated), it actually doesn't matter that we're only processing one of those ref_paths. The `build_reference_dicts` just assembles the resources themselves that are connected to the anchor in an aggregated fashion--the `ref_path` is just used to figure out what the connection to the anchor is. This means that when we actually get to `tabulate`, because we're iterating back over the _schema columns_ rather than the ref_paths, the proper value can still be looked up and extracted. On the other hand, if we were to implement this change and make the reverse directions hold a list of ref_paths, the `build_reference_dicts` function would interpret those as two separate resources, which would lead to every resource linked to anchor being added to the reference dictionary twice. This in turn would cause the tabulate function to double report every value that was a reverse reference.

To back up everything that I'm saying, I just modified the test cases I have locally and ran them through. I asked for another reverse reference Observation (to make sure nothing was excluded or double counted) in the schema, and then without changing algorithm logic at all, checked the output. The code successfully added the new column with the new information from the additional observations without double counting anything. So, I think this case is already handled and not a concern with the algorithmm.